runtime/gc_modron_standard/StandardAccessBarrier.cpp

```diff
@@ -933,3 +915,164 @@ MM_StandardAccessBarrier::preObjectRead(J9VMThread *vmThread, J9Object *srcObjec
 }
 #endif
 
+void
+MM_StandardAccessBarrier::forcedToFinalizableObject(J9VMThread* vmThread, J9Object* object)
+{
+	if (!_extensions->usingSATBBarrier()) {
+		MM_ObjectAccessBarrier::forcedToFinalizableObject(vmThread, object);
+		return;
+	}
+
+	MM_EnvironmentBase* env = MM_EnvironmentBase::getEnvironment(vmThread->omrVMThread);
+	if (_extensions->isSATBBarrierActive()) {
+		rememberObjectToRescan(env, object);
+	}
+}
+
+/**
+ * Override of referenceGet.  When the collector is tracing, it makes any gotten object "grey" to ensure
+ * that it is eventually traced.
+ *
+ * @param refObject the SoftReference or WeakReference object on which get() is being called.
+ *	This barrier must not be called for PhantomReferences.  The parameter must not be NULL.
+ */
+J9Object *
+MM_StandardAccessBarrier::referenceGet(J9VMThread *vmThread, J9Object *refObject)
+{
+	if (!_extensions->usingSATBBarrier()) {
+		return MM_ObjectAccessBarrier::referenceGet(vmThread, refObject);
+	}
+
+	MM_ParallelGlobalGC *globalCollector = (MM_ParallelGlobalGC *)_extensions->getGlobalCollector();
+	MM_EnvironmentBase* env = MM_EnvironmentBase::getEnvironment(vmThread->omrVMThread);
+
+	UDATA offset = J9VMJAVALANGREFREFERENCE_REFERENT_OFFSET(vmThread);
+	J9Object *referent = mixedObjectReadObject(vmThread, refObject, offset, false);
+
```

Comment:
part up to this point can probably be simplified, since base referenceGet, is doing effectively the same as mixedObjectReadObject

Reply:
I can't seem to find the define for `J9VMJAVALANGREFREFERENCE_REFERENT_OFFSET`

Reply:
The following are of interest for this:
```C++
#define J9VMJAVALANGREFREFERENCE_REFERENT_VM(javaVM, object) ((void)0, J9OBJECT_OBJECT_LOAD_VM(javaVM, object, J9VMJAVALANGREFREFERENCE_REFERENT_OFFSET_VM(javaVM)))
```
```C++
#if defined (J9VM_GC_ALWAYS_CALL_OBJECT_ACCESS_BARRIER)
#define J9OBJECT_OBJECT_LOAD_VM(javaVM, object, offset) ((j9object_t)javaVM->memoryManagerFunctions->j9gc_objaccess_mixedObjectReadObject(J9JAVAVM_VMTHREAD(javaVM), (j9object_t)(object), (offset), 0))
...
#else /* J9VM_GC_ALWAYS_CALL_OBJECT_ACCESS_BARRIER */
...
#define J9OBJECT_OBJECT_LOAD_VM(javaVM, object, offset) \
	(J9OBJECT__PRE_OBJECT_LOAD_VM(javaVM, object, offset), \
	(J9JAVAVM_COMPRESS_OBJECT_REFERENCES(javaVM) \
		? (j9object_t)J9_CONVERT_POINTER_FROM_TOKEN_VM__(javaVM, *(U_32*)((U_8*)(object) + (offset))) \
		: (j9object_t)*(UDATA*)((U_8*)(object) + (offset))))
```

Reply:
Not sure if there are any performance implications to making this change. Currently, we take path 
`referenceGet` -> `J9VMJAVALANGREFREFERENCE_REFERENT_VM` which ends up doing:
```
	..... ? (j9object_t)J9_CONVERT_POINTER_FROM_TOKEN_VM__(javaVM, *(U_32*)((U_8*)(object) + (offset))) \
		: (j9object_t)*(UDATA*)((U_8*)(object) + (offset))))
```

With this change we have to go through a few different barrier methods,  `mixedObjectReadObject` -> `readObjectImpl` and some other barrier methods like `postObjectRead` and `preObjectRead` + `protectIfVolatileBefore/After`

```C+++
J9Object *
MM_ObjectAccessBarrier::mixedObjectReadObject(J9VMThread *vmThread, J9Object *srcObject, UDATA srcOffset, bool isVolatile)
{
	fj9object_t *actualAddress = J9OAB_MIXEDOBJECT_EA(srcObject, srcOffset, fj9object_t);
	J9Object *result = NULL;

	if (preObjectRead(vmThread, srcObject, actualAddress)) {
		protectIfVolatileBefore(vmThread, isVolatile, true, false);
		result = readObjectImpl(vmThread, srcObject, actualAddress, isVolatile);
		protectIfVolatileAfter(vmThread, isVolatile, true, false);

		if (!postObjectRead(vmThread, srcObject, actualAddress)) {
			result = NULL;
		}
	}
	
	/* This must always be called to massage the return value */
	return result;
}
```

Reply:
Changes made